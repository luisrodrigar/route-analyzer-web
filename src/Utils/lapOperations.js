import randomColor from 'randomcolor'; 

  // Unused because It delegates calculating the zoom to fitBound in RouteMapComponent (zoomToMarkers)  
  export function getCenterLaps(laps){
    let resumeCenterByLap = laps.map(lap=>{
        return getCenterOfALap(lap);
      });
    const lat = resumeCenterByLap.reduce((total, center)=>{
      return total = total + parseFloat(center.lat);
    }, 0.0)/laps.length;
    const lng = resumeCenterByLap.reduce((total, center)=>{
      return total = total + parseFloat(center.lng);
    }, 0.0)/laps.length;

    return {
      lat,
      lng
    }
  }

  export function getLapsTrackPoints(laps){
    return laps.map(lap => {
        return {
              index:lap.index,
              tracks:getLapTrackPoint(lap),
              color:randomColor(),
              startTime: lap.startTime,
              totalTime: lap.totalTimeSeconds,
              distance: lap.distanceMeters,
              maxSpeed: lap.maximunSpeed,
              avgSpeed: lap.averageSpeed,
              avgBpm: lap.averageHearRate,
              maxBpm: lap.maximunHeartRate,
              cal: lap.calories,
              intensity: lap.intensity
        }
    });
  }

  function getLapTracksWithPosition(lap){
    return lap.tracks
      .filter(track => {
        if(track && track.position)
          return track;
        else return null;
      });
  }

  function getLapTrackPoint(lap){
    return getLapTracksWithPosition(lap).map(trackPoint => {
        var lat = parseFloat(trackPoint.position.latitudeDegrees);
        var lng = parseFloat(trackPoint.position.longitudeDegrees);
        var alt = parseFloat(trackPoint.altitudeMeters);
        var speed = parseFloat(trackPoint.speed);
        var date = parseInt(trackPoint.date,10);
        var dist = parseFloat(trackPoint.distanceMeters);
        var bpm = parseInt(trackPoint.heartRateBpm,10);
        var index = parseInt(trackPoint.index,10);
        return {
          index,
          date,
          position:{
            lat,
            lng
          },
          alt,
          speed,
          dist,
          bpm
        };
    });
  }


  function getCenterOfALap(lap){
    const tracksPositionNotNull = getLapTracksWithPosition(lap);
    const lat = tracksPositionNotNull
      .map(track =>{
        return parseFloat(track.position.latitudeDegrees);
      }).reduce(
        (total,lat) => {return total+lat;}, 0.0
      )/tracksPositionNotNull.length;

    const lng = tracksPositionNotNull
      .map(track=>{
        return parseFloat(track.position.longitudeDegrees);
      }).reduce(
        (total,lng)=> {return total+lng;},0.0
      )/tracksPositionNotNull.length;

    return { 
      lat,
      lng
    } 
  }