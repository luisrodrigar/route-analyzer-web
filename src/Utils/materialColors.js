import {red, pink, purple, deepPurple, indigo, blue,
        lightBlue, cyan, teal, green, lightGreen, 
        lime, yellow, amber, orange, deepOrange, brown, grey, blueGrey  } from 'material-ui/colors';

const mainColors = [
  'Red',
  'Pink',
  'Purple',
  'Deep Purple',
  'Indigo',
  'Blue',
  'Light Blue',
  'Cyan',
  'Teal',
  'Green',
  'Light Green',
  'Lime',
  'Yellow',
  'Amber',
  'Orange',
  'Deep Orange',
  'Brown', 
  'Grey',
  'Blue Grey'
];
const lightPalette = [200];
const darkPalette = [500,900];

function getColorObject(name){
  switch(name){
    case mainColors[0]:
    return red;
    case mainColors[1]:
    return pink;
    case mainColors[2]:
    return purple;
    case mainColors[3]:
    return deepPurple;
    case mainColors[4]:
    return indigo;
    case mainColors[5]:
    return blue;
    case mainColors[6]:
    return lightBlue;
    case mainColors[7]:
    return cyan;
    case mainColors[8]:
    return teal;
    case mainColors[9]:
    return green;
    case mainColors[10]:
    return lightGreen;
    case mainColors[11]:
    return lime;
    case mainColors[12]:
    return yellow;
    case mainColors[13]:
    return amber;
    case mainColors[14]:
    return orange;
    case mainColors[15]:
    return deepOrange;
    case mainColors[16]:
    return brown;
    case mainColors[17]:
    return grey;
    case mainColors[18]:
    return blueGrey;
    default:
    return getColorObject(mainColors[Math.floor(Math.random()*mainColors.length)]);
  }
}

function getColor(){
  const colorDark = darkPalette[Math.floor(Math.random()*darkPalette.length)];
  const colorLight = lightPalette[Math.floor(Math.random()*lightPalette.length)];
  const colorName = mainColors[Math.floor(Math.random()*mainColors.length)];

  const colorObject = getColorObject(colorName);

  let colorValue = colorObject[colorDark];
  let colorLightValue = colorObject[colorLight];

  return [colorValue,colorLightValue];
}

export function getColorRandom(numColors, currentColors){
  let arrayColors = [];
  let color = null;
  for(let i=0; i<numColors; i++){
    if(numColors<(mainColors.length * darkPalette.length)){
      do{
        color = getColor();
      }while(currentColors.map(x=>x.join('')).indexOf(color.join(''))>=0);
      currentColors.push(color);
    }else{
      color = getColor();
    }
    arrayColors.push(color);
  }
  return arrayColors;
}