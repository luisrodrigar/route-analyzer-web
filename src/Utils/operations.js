import {getColorRandom} from './colors';
import * as date        from 'datejs';

export function getLapsTrackPoints(laps){
  return laps.map(lap => {
      return {
            index:lap.index,
            tracks:getLapTrackPoint(lap),
            color:lap.color,
            lightColor: lap.lightColor,
            startTime: lap.startTime ?  parseDate(lap.startTime).getTime() : null,
            totalTime: lap.totalTimeSeconds,
            distance: lap.distanceMeters,
            maxSpeed: lap.maximumSpeed,
            avgSpeed: lap.averageSpeed,
            avgBpm: lap.averageHearRate,
            maxBpm: lap.maximumHeartRate,
            cal: lap.calories,
            intensity: lap.intensity
      }
  });
}

function getLapTrackPoint(lap){
  return getLapTracksWithPosition(lap).map(trackPoint => {
      const lat = parseFloat(trackPoint.position.latitudeDegrees);
      const lng = parseFloat(trackPoint.position.longitudeDegrees);
      const position = {lat, lng};
      const alt = parseFloat(trackPoint.altitudeMeters);
      const speed = parseFloat(trackPoint.speed);
      const date = trackPoint.date ? parseDate(trackPoint.date).getTime() : null;
      const dist = parseFloat(trackPoint.distanceMeters);
      const bpm = parseInt(trackPoint.heartRateBpm,10);
      const index = parseInt(trackPoint.index,10);
      return {
        index,
        date,
        position,
        alt,
        speed,
        dist,
        bpm
      };
  });
}

function parseDate(dateString){
  return Date.parse(dateString);
}

export function setLapColors(laps){
  let currentColors = laps
      .filter(lap => lap.color && lap.lightColor)
      .map(lap => [lap.color,lap.lightColor]);
  let numLapsWithoutColor = laps
      .filter(lap=> !lap.color && !lap.lightColor)
      .length;
  const colors = getColorRandom(numLapsWithoutColor, currentColors);
  laps.filter(lap => !lap.color && !lap.lightColor)
      .forEach((lap, index) => {
        lap.color = colors[index][0];
        lap.lightColor = colors[index][1];
      });
}

export function getLapsSpeed(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.lightColor,
      label: "Lap " + lap.index,
      stroke: 2,
      tracks:lap.tracks.map(track=>[track.date, track.speed])
    }
  });
}

export function getLapsHeartRate(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.lightColor,
      label: "Lap " + lap.index,
      tracks: lap.tracks.map(track => [track.date, track.bpm])
    }
  });
}

export function getLapsElevations(laps){
  return laps.map(lap=> {
    return {
      index: lap.index,
      color: lap.color,
      label: "Lap " + lap.index,
      tracks: lap.tracks.map(track => [track.date, track.alt])
    }
  });
}

export function calculateAvgValues(laps, avgMethod) {
    return laps
        .map(lap => [getAvgDate(lap.tracks), avgMethod(lap)]);
}

function getAvgDate(tracks) {
    return (tracks[0].date + tracks[tracks.length-1].date) / 2;
}

function calculateAvgSingleLap(laps, avgArray, avgMethod) {
    let avgDate = avgArray[0][0];
    let initDate = laps[0].tracks[0].date;
    let endDate = laps[0].tracks[(laps[0].tracks.length)-1].date;
    avgArray = [];
    avgArray.push([initDate, avgMethod(laps[0])]);
    avgArray.push([avgDate, avgMethod(laps[0])]);
    avgArray.push([endDate, avgMethod(laps[0])]);
    return avgArray;
}

export function getAvgBpm(laps){
    let avgBpmMethod = function(lap) {
        return lap.avgBpm;
    };
    let avgBpm = calculateAvgValues(laps, avgBpmMethod);
    return laps.length > 1 ? avgBpm : calculateAvgSingleLap(laps, avgBpm, avgBpmMethod);
}

export function getAvgSpeed(laps){
    let avgSpeedMethod = function(lap) {
        return lap.avgSpeed;
    };
    let avgSpeed = calculateAvgValues(laps, avgSpeedMethod);
    return laps.length > 1 ? avgSpeed : calculateAvgSingleLap(laps, avgSpeed, avgSpeedMethod);
}

export function getHeartRateData(laps) {
  return laps.flatMap(lap =>
      lap.tracks.map(track => [track.date,track.bpm]));
}

export function getElevationData(laps) {
  return laps.flatMap(lap =>
    lap.tracks.map(track => [track.date,track.alt]));
}

export function getSpeedData(laps){
  return laps.flatMap(lap =>
      lap.tracks.map(track => [track.date,track.speed]));
}

function getLapTracksWithPosition(lap){
  return lap.tracks
    .filter(track => {
      if(track && track.position)
        return track;
      else return null;
    });
}

export function getNearestPosition(laps,position){
  let indexLap =null, indexTrackpoint = null, minDistance = Number.POSITIVE_INFINITY;
  laps.forEach((lap, eachIndexLap)=>{
    lap.tracks.forEach((track,eachIndexTrackpoint)=>{
      let eachPosition = {
        lat: track.position.lat,
        lng: track.position.lng,
      };
      let currentMin = getDistanceBetweenPoints(eachPosition,position);
      if(currentMin<minDistance){
        indexTrackpoint = eachIndexTrackpoint;
        indexLap = eachIndexLap;
        minDistance = currentMin;
      }
    })
  })
  return {
    indexLap,
    indexTrackpoint
  }
}

function getDistanceBetweenPoints(point1, point2){
  // Convert degrees to radians
  let latP1 = degrees2Radians(point1.lat), lngP1 = degrees2Radians(point1.lng);
  let latP2 = degrees2Radians(point2.lat), lngP2 = degrees2Radians(point2.lng);

  // Radius of earth in meters
  let earthRadiusMeters = 6378100;

  // Point P 
  let rho1 = earthRadiusMeters * Math.cos(latP1);
  let z1 = earthRadiusMeters * Math.sin(latP1);
  let x1 = rho1 * Math.cos(lngP1);
  let y1 = rho1 * Math.sin(lngP1);

  // Point Q
  let rho2 = earthRadiusMeters * Math.cos(latP2);
  let z2 = earthRadiusMeters * Math.sin(latP2);
  let x2 = rho2 * Math.cos(lngP2);
  let y2 = rho2 * Math.sin(lngP2);

  // Dot product
  let dot = (x1 * x2 + y1 * y2 + z1 * z2);
  let cosTheta = dot / (Math.pow(earthRadiusMeters, 2));

  let theta = Math.acos(cosTheta);

  return earthRadiusMeters * theta;

}

function degrees2Radians(degree){
  return degree * Math.PI / 180.0;
}
