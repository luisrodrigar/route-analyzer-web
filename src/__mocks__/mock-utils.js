import {
    getElevationsData,
    getElevationsLaps,
    getHeartRateAvgValues,
    getHeartRateData,
    getHeartRateLaps, getSpeedAvgData, getSpeedData, getSpeedLaps
} from "../__test__/test-utilities";

export function mockHeartRateMethods(operations) {
    operations.getLapsHeartRate.mockImplementationOnce(() => getHeartRateLaps());
    operations.getHeartRateData.mockImplementationOnce(() => getHeartRateData());
    operations.getAvgBpm.mockImplementationOnce(() => getHeartRateAvgValues());
    return operations;
}

export function mockElevationsMethods(operations) {
    operations.getElevationData.mockImplementationOnce(() => getElevationsData());
    operations.getLapsElevations.mockImplementationOnce(() => getElevationsLaps());
    return operations;
}

export function mockSpeedMethods(operations) {
    operations.getLapsSpeed.mockImplementationOnce(() => getSpeedLaps());
    operations.getSpeedData.mockImplementationOnce(() => getSpeedData());
    operations.getAvgSpeed.mockImplementationOnce(() => getSpeedAvgData());
    return operations;
}

