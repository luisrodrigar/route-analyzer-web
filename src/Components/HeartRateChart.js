import React    from 'react';
import sizeMe from 'react-sizeme';
import Grid from '@material-ui/core/Grid';
import BarChart from './Charts/BarChart'
import {getLapsHeartRate, getHeartRateData, getAvgBpm} from '../Utils/operations';
import { connect } from "react-redux";
import { updateTrackpoint } from "../actions/index";

// Using a react-sizeme library to get width and height values => Handling on resize event
const MySizeGrid = sizeMe({ monitorHeight: true })(Grid);

const mapStateToProps = state => {
  return {
    currentTrackpoint:  state.container.currentTrackpoint,
    laps:               state.container.laps
  };
};

const mapDispatchToProps = dispatch => {
  return {
    updateTrackpoint: (idLap, idTrackpoint) => dispatch(updateTrackpoint(idLap, idTrackpoint))
  };
};

export class HeartRateChart extends React.Component {

  constructor(props){
    super(props);
    this.handleChangeSizeContent = this.handleChangeSizeContent.bind(this);
    let lapsHeartRate = getLapsHeartRate(this.props.laps);
    let heartRates = getHeartRateData(this.props.laps);
    let avgLine = getAvgBpm(this.props.laps);
    this.state = {
      bpms:   heartRates,
      laps:   lapsHeartRate,
      avg: avgLine,
      width:  1400,
      height: 200,
    };
  }

  componentDidUpdate(prevProps, prevState, prevContext){
    if(prevProps.laps !== this.props.laps){
      let chart = this.chartGrid;
      let heartRates = getHeartRateData(this.props.laps);
      let lapsHeartRate = getLapsHeartRate(this.props.laps);
      let avgLine = getAvgBpm(this.props.laps);
      this.setState({
        bpms: heartRates,
        laps: lapsHeartRate,
        avg: avgLine,
        width:chart.state.width,
        height:chart.state.height,
      });
    }
  }

  handleChangeSizeContent(){
    let chart = this.chartGrid;
    this.setState({
      width:chart.state.width,
      height:chart.state.height,
    });
  }

  componentDidMount(){
    window.addEventListener("resize", this.handleChangeSizeContent);
  }

  componentWillUnmount(){
    window.removeEventListener("resize", this.handleChangeSizeContent);
  }

  render(){
		return(
		    <MySizeGrid
                ref={ref=>this.chartGrid = ref} item
                xs={12} style={{maxHeight:'200px', minWidth:'500px'}}>
              <BarChart   textYAxis={this.props.yTitle}
                          textXAxis={this.props.xTitle}
                          track={this.props.currentTrackpoint}
                          handleMouseOver={this.props.updateTrackpoint}
                          data={this.state.bpms}
                          laps={this.state.laps}
                          dataLine={this.state.avg}
                          width = {this.state.width}
                          height = {this.state.height}
                          legend={'bpm'}
                          padding = {40}
                          ticks={5}/>
		    </MySizeGrid>
        );
  }

}

export default connect(mapStateToProps,mapDispatchToProps)(HeartRateChart);
