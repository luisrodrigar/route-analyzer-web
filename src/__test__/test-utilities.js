import * as operations from "../Utils/operations";

export function getLaps() {
    return [
        {
            index: 1,
            startTime: "2018-12-21T14:38:07Z",
            totalTimeSeconds: 10,
            maximumSpeed: 4.727421976808574,
            averageSpeed: 2.9996712051,
            averageHearRate: 94,
            maximumHeartRate: 98,
            calories: 67,
            intensity: "LOW",
            color: '#ff320b',
            lightColor: '#cccc80',
            distanceMeters: null,
            tracks: [
                {
                    date: "2018-12-21T14:38:07Z",
                    index: 1,
                    position: {
                        latitudeDegrees: 42.548370361328125,
                        longitudeDegrees: -6.5837578773498535
                    },
                    altitudeMeters: 556.4249877929688,
                    speed: 1.271920433274412,
                    heartRateBpm: 90,
                    distanceMeters: 0
                },
                {
                    date: "2018-12-21T14:38:17Z",
                    index: 2,
                    position: {
                        latitudeDegrees: 42.54838943481445,
                        longitudeDegrees: -6.583323001861572
                    },
                    altitudeMeters: 556.6911010742188,
                    speed: 4.727421976808574,
                    heartRateBpm: 98,
                    distanceMeters: 20
                }
            ]
        },
        {
            index: 2,
            startTime: "2018-12-21T14:38:27Z",
            totalTimeSeconds: 10,
            maximumSpeed: 4.533650591446289,
            averageSpeed: 3.1588017983,
            averageHearRate: 115,
            maximumHeartRate: 120,
            calories: 145,
            intensity: "HIGH",
            color: '#ffbb97',
            lightColor: '#00aa80',
            distanceMeters: null,
            tracks: [
                {
                    date: "2018-12-21T14:38:27Z",
                    index: 1,
                    position: {
                        latitudeDegrees: 42.54792022705078,
                        longitudeDegrees: -6.5830583572387695
                    },
                    altitudeMeters: 553.3866577148438,
                    speed: 1.7839530051752528,
                    heartRateBpm: 110,
                    distanceMeters: 0
                },
                {
                    date: "2018-12-21T14:38:37Z",
                    index: 2,
                    position: {
                        latitudeDegrees: 42.54783248901367,
                        longitudeDegrees: -6.58258581161499
                    },
                    altitudeMeters: 551.2217407226562,
                    speed: 4.533650591446289,
                    heartRateBpm: 120,
                    distanceMeters: 57
                }
            ]
        },
        {
            index: 3,
            totalTimeSeconds: 0,
            startTime: "2018-12-21T14:38:47Z",
            maximumSpeed: 2.179150563690673,
            averageSpeed: 2.179150563690673,
            averageHearRate: 135,
            maximumHeartRate: 135,
            calories: 89,
            intensity: "LOW",
            color: '#0214ff',
            lightColor: '#caac80',
            distanceMeters: 0,
            tracks: [
                {
                    date: "2018-12-21T14:38:47Z",
                    index: 1,
                    position: {
                        latitudeDegrees: 42.546897888183594,
                        longitudeDegrees: -6.582358360290527
                    },
                    altitudeMeters: 542.0283203125,
                    speed: 2.179150563690673,
                    heartRateBpm: 135,
                    distanceMeters: 0
                }
            ]
        }
    ];
}

export function getHeartRateData() {
    return [
        [1545403087000, 90],
        [1545403097000, 98],
        [1545403107000, 110],
        [1545403117000, 120],
        [1545403127000, 135]
    ];
}

export function getHeartRateLaps() {
    return [
        {
            index: 1,
            color: '#cccc80',
            label: "Lap 1",
            tracks: [
                [1545403087000, 90],
                [1545403097000, 98],
            ]
        },
        {
            index: 2,
            color: '#00aa80',
            label: "Lap 2",
            tracks: [
                [1545403107000, 110],
                [1545403117000, 120]
            ]
        },
        {
            index: 3,
            color: '#caac80',
            label: "Lap 3",
            tracks: [
                [1545403127000, 135]
            ]
        }
    ];
}

export function getHeartRateAvgValues() {
    return [
        [1545403092000, 94],
        [1545403112000, 115],
        [1545403127000, 135]
    ];
}

export function getCurrentTrackPoint() {
    return {
        date: "2018-12-21T14:38:17",
        index: 2,
        position: {
            latitudeDegrees: 42.54838943481445,
            longitudeDegrees: -6.583323001861572
        },
        altitudeMeters: 556.6911010742188,
        speed: 4.727421976808574,
        heartRateBpm: 98
    };
}

export function getElevationsData() {
    return [
        [1545403087000, 556.4249877929688],
        [1545403097000, 556.6911010742188],
        [1545403107000, 553.3866577148438],
        [1545403117000, 551.2217407226562],
        [1545403127000, 542.0283203125]
    ];
}

export function getElevationsLaps() {
    return [
        {
            index: 1,
            color: '#ff320b',
            label: "Lap 1",
            tracks: [
                [1545403087000, 556.4249877929688],
                [1545403097000, 556.6911010742188],
            ]
        },
        {
            index: 2,
            color: '#ffbb97',
            label: "Lap 2",
            tracks: [
                [1545403107000, 553.3866577148438],
                [1545403117000, 551.2217407226562]
            ]
        },
        {
            index: 3,
            color: '#0214ff',
            label: "Lap 3",
            tracks: [
                [1545403127000, 542.0283203125]
            ]
        }
    ];
}

export function getSpeedData() {
    return [
        [1545403087000, 1.271920433274412],
        [1545403097000, 4.727421976808574],
        [1545403107000, 1.7839530051752528],
        [1545403117000, 4.533650591446289],
        [1545403127000, 2.179150563690673]
    ];
}

export function getSpeedLaps() {
    return [
        {
            index: 1,
            color: '#cccc80',
            label: "Lap 1",
            tracks: [
                [1545403087000, 1.271920433274412],
                [1545403097000, 4.727421976808574],
            ]
        },
        {
            index: 2,
            color: '#00aa80',
            label: "Lap 2",
            tracks: [
                [1545403107000, 1.7839530051752528],
                [1545403117000, 4.533650591446289]
            ]
        },
        {
            index: 3,
            color: '#caac80',
            label: "Lap 3",
            tracks: [
                [1545403127000, 2.179150563690673]
            ]
        }
    ];
}

export function getExpectedSpeedLaps() {
    let speedLaps = getSpeedLaps();
    speedLaps[0].stroke = 2;
    speedLaps[1].stroke = 2;
    speedLaps[2].stroke = 2;
    return speedLaps;
}

export function getSpeedAvgData() {
    return [
        [1545403092000, 2.9996712051],
        [1545403112000, 3.1588017983],
        [1545403127000, 2.179150563690673]
    ];
}

export function getRouteAnalyzerWebLaps() {
    return [
        {
            index:1,
            tracks:[
                {
                    index: 1,
                    position:{
                        lat: 42.548370361328125,
                        lng: -6.5837578773498535
                    },
                    alt: 556.4249877929688,
                    speed: 1.271920433274412,
                    date: 1545403087000,
                    dist: 0,
                    bpm: 90
                },
                {
                    index: 2,
                    position:{
                        lat: 42.54838943481445,
                        lng: -6.583323001861572
                    },
                    alt: 556.6911010742188,
                    speed: 4.727421976808574,
                    date: 1545403097000,
                    dist: 20,
                    bpm: 98
                }
            ],
            color: '#ff320b',
            lightColor: '#cccc80',
            startTime: 1545403087000,
            totalTime: 10,
            distance: null,
            maxSpeed: 4.727421976808574,
            avgSpeed: 2.9996712051,
            avgBpm: 94,
            maxBpm: 98,
            cal: 67,
            intensity: "LOW"
        },
        {
            index:2,
            tracks:[
                {
                    index: 1,
                    position:{
                        lat: 42.54792022705078,
                        lng: -6.5830583572387695
                    },
                    alt: 553.3866577148438,
                    speed: 1.7839530051752528,
                    date: 1545403107000,
                    dist: 0,
                    bpm: 110
                },
                {
                    index: 2,
                    position: {
                        lat: 42.54783248901367,
                        lng: -6.58258581161499
                    },
                    alt: 551.2217407226562,
                    speed: 4.533650591446289,
                    date: 1545403117000,
                    dist: 57,
                    bpm: 120
                }
            ],
            color:'#ffbb97',
            lightColor: '#00aa80',
            startTime: 1545403107000,
            totalTime: 10,
            distance: null,
            maxSpeed: 4.533650591446289,
            avgSpeed: 3.1588017983,
            avgBpm: 115,
            maxBpm: 120,
            cal: 145,
            intensity: "HIGH"
        },
        {
            index:3,
            tracks:[
                {
                    index: 1,
                    date: 1545403127000,
                    position:{
                        lat: 42.546897888183594,
                        lng: -6.582358360290527
                    },
                    alt: 542.0283203125,
                    speed: 2.179150563690673,
                    dist: 0,
                    bpm: 135
                }
            ],
            color: '#0214ff',
            lightColor: '#caac80',
            startTime: 1545403127000,
            totalTime: 0,
            distance: 0,
            maxSpeed: 2.179150563690673,
            avgSpeed: 2.179150563690673,
            avgBpm: 135,
            maxBpm: 135,
            cal: 89,
            intensity: "LOW"
        }
    ];
}

export function lapsWithoutColors() {
    return [
        {
            index:1,
            tracks: [],
            color: null,
            lightColor: null,
            startTime: 1545403087000,
            totalTime: 10
        },
        {
            index:2,
            tracks: [],
            color:'#ffbb97',
            lightColor: '#00aa80',
            startTime: 1545403107000,
            totalTime: 10
        },
        {
            index:3,
            tracks: [],
            color: null,
            lightColor: null,
            startTime: 1545403127000,
            totalTime: 0
        }
    ];
}

export function getArrayColors() {
    return [['#ff320b', '#cc7e6c'],['#0214ff', '#8798ca']];
}

export function expectedLapsWithColor() {
    return [
        {
            index:1,
            tracks: [],
            color: '#ff320b',
            lightColor: '#cc7e6c',
            startTime: 1545403087000,
            totalTime: 10
        },
        {
            index:2,
            tracks: [],
            color:'#ffbb97',
            lightColor: '#00aa80',
            startTime: 1545403107000,
            totalTime: 10
        },
        {
            index:3,
            tracks: [],
            color: '#0214ff',
            lightColor: '#8798ca',
            startTime: 1545403127000,
            totalTime: 0
        }
    ];
}

export function getRouteAnalyzerWebOneLap() {
    return [
        {
            index:1,
            tracks:[
                {
                    index: 1,
                    position:{
                        lat: 42.548370361328125,
                        lng: -6.5837578773498535
                    },
                    alt: 556.4249877929688,
                    speed: 1.271920433274412,
                    date: 1545403087000,
                    dist: 0,
                    bpm: 90
                },
                {
                    index: 2,
                    position:{
                        lat: 42.54838943481445,
                        lng: -6.583323001861572
                    },
                    alt: 556.6911010742188,
                    speed: 4.727421976808574,
                    date: 1545403097000,
                    dist: 20,
                    bpm: 98
                }
            ],
            color: '#ff320b',
            lightColor: '#cccc80',
            startTime: 1545403087000,
            totalTime: 10,
            distance: null,
            maxSpeed: 4.727421976808574,
            avgSpeed: 2.9996712051,
            avgBpm: 94,
            maxBpm: 98,
            cal: 67,
            intensity: "LOW"
        }
    ];
}

export function getSpeedAvgDataOneLap() {
    return [
        [1545403087000, 2.9996712051],
        [1545403092000, 2.9996712051],
        [1545403097000, 2.9996712051]
    ];
}

export function getHeartRateAvgValuesOneLap() {
    return [
        [1545403087000, 94],
        [1545403092000, 94],
        [1545403097000, 94]
    ];
}
