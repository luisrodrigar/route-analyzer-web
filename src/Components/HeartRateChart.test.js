import React from "react";
import {HeartRateChart} from "./HeartRateChart";
import sizeMe from 'react-sizeme';
import {mount, shallow} from "enzyme";
import Grid from '@material-ui/core/Grid';
import toJson from "enzyme-to-json";
import {getCurrentTrackPoint, getLaps} from "../__test__/test-utilities";
import {mockHeartRateMethods} from "../__mocks__/mock-utils";
import * as operations from '../Utils/operations';

jest.mock('../Utils/operations');
jest.mock('../actions/index');

describe("Heart Rate Chart Test", () => {
    let updateTrackPointMock;
    let currentTrackPoint;
    let laps;
    let component;
    let props;
    beforeAll(() => {
        currentTrackPoint = getCurrentTrackPoint();
        laps = getLaps();
        props = {
            currentTrackpoint: currentTrackPoint,
            laps: laps,
        };
    });
    beforeEach( () => {
        mockHeartRateMethods(operations);
        updateTrackPointMock = jest.fn();
    });
    afterEach( () => {
        jest.resetAllMocks();
        component.unmount();
    });
    it("Checking the heart rate snapshot",async () => {
        component = shallow(<HeartRateChart />);
        expect(toJson(component)).toMatchSnapshot();
    });
    it("Checking the resize event on window heart rate.",async () => {
        const mockSizeGrid = sizeMe({monitorHeight: true})(Grid);
        const widthMockValue = 15;
        const heightMockValue = 10;
        mockSizeGrid.state = {
            width: widthMockValue,
            height: heightMockValue
        };
        component = mount(<HeartRateChart props={props}
                                           yTitle={'Heart Rate (bpm)'}
                                           xTitle={'Time (hh:mm:ss)'}
                                           updateTrackpoint={updateTrackPointMock}/>);
        const instance = component.instance();
        instance.chartGrid = mockSizeGrid;
        window.dispatchEvent(new Event('resize'));
        expect(instance.state.width).toEqual(widthMockValue);
        expect(instance.state.height).toEqual(heightMockValue);
    });
});
