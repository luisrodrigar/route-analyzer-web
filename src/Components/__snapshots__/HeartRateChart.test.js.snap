// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heart Rate Chart Test Checking the heart rate snapshot 1`] = `
<SizeMe(WithStyles(Grid))
  item={true}
  style={
    Object {
      "maxHeight": "200px",
      "minWidth": "500px",
    }
  }
  xs={12}
>
  <BarChart
    data={
      Array [
        Array [
          1545403087000,
          90,
        ],
        Array [
          1545403097000,
          98,
        ],
        Array [
          1545403107000,
          110,
        ],
        Array [
          1545403117000,
          120,
        ],
        Array [
          1545403127000,
          135,
        ],
      ]
    }
    dataLine={
      Array [
        Array [
          1545403092000,
          94,
        ],
        Array [
          1545403112000,
          115,
        ],
        Array [
          1545403127000,
          135,
        ],
      ]
    }
    height={200}
    laps={
      Array [
        Object {
          "color": "#cccc80",
          "index": 1,
          "label": "Lap 1",
          "tracks": Array [
            Array [
              1545403087000,
              90,
            ],
            Array [
              1545403097000,
              98,
            ],
          ],
        },
        Object {
          "color": "#00aa80",
          "index": 2,
          "label": "Lap 2",
          "tracks": Array [
            Array [
              1545403107000,
              110,
            ],
            Array [
              1545403117000,
              120,
            ],
          ],
        },
        Object {
          "color": "#caac80",
          "index": 3,
          "label": "Lap 3",
          "tracks": Array [
            Array [
              1545403127000,
              135,
            ],
          ],
        },
      ]
    }
    legend="bpm"
    padding={40}
    ticks={5}
    width={1400}
  />
</SizeMe(WithStyles(Grid))>
`;
